### DISCUSSION TITLE
Exploring the Discrepancies in UI Testing (Feature Testing) Speed Between iOS and Android Platforms

### DISCUSSION THREAD

**Discussion by novinfard** [MobileNativeFoundation/discussions#198] [General]
The similar UI test case for iOS takes around 10 times more than Android counterpart. 

How can the significant difference in UI testing speed between iOS and Android be explained, and what strategies can be used to address this challenge?

More specifically, have you encountered a similar discrepancy in test case execution times between the two platforms, and if so, how have you tackled it? Additionally, what are the contributing factors to this difference, taking into account the underlying testing frameworks - XCTest for iOS and Espresso for Android?

It is worth noting that in one experiment, the initial launch of the app for iOS UI testing alone took approximately 4 seconds, which may be indicative of further factors impacting the overall testing speed on this platform.

**Comment by prolificcoder** [comment]
Espresso is a gray box testing tool meaning that it knows the internals of the app (not as much as white box/unit tests) but enough to make the [synchronization](https://developer.android.com/training/testing/espresso) between the app and the testing framework faster. Espresso can easily invoke arbitrary activities, intents, access helper methods, strings.xml etc. One downside is that you can't easily access the sandbox outside of your app. 

XCUITesting is a E2E testing tool which doesn't have any such mechanisms built into it. It's not easy to access app code/helpers that you created for unit tests or for the app. 

You can't really compare these two frameworks. Test automation on Android is much more pleasurable, reliable and faster. 
Espresso's predecessor UIAutomator2, is an E2E tool and is similar to XCUITesting and you could expect similar timings from it. 

One iOS alternative that was pursued by google and similar to espresso is [EarlGray](https://github.com/google/EarlGrey) but it looks abandoned now. 

As a developer that have spend over 8 years on automating both platforms, my advice is to shift the testing down as much as possible. i.e more unit tests than UI tests. And mock all the UI tests, do not interact with real network data, either get it from raw JSON or data objects but do not make real API calls. 

Let me know if you have any further questions, I enjoy this topic and can go deeper if you have specific questions.

**Reply by novinfard** [reply]
Thank you for your helpful response. 🙏

> One iOS alternative that was pursued by google and similar to espresso is [EarlGray](https://github.com/google/EarlGrey) but it looks abandoned now.

It's good to see if we can find any similar gray box tooling on iOS platform. In EarlGray v2 version they changed to use [XCUITest](https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/testing_with_xcode/chapters/09-ui_testing.html) which may result in the same speed issue. The current alternative could be [Detox](https://github.com/wix/Detox), but glad to hear that we have more options here.

> And mock all the UI tests, do not interact with real network data, either get it from raw JSON or data objects but do not make real API calls.
We mock all the network requests, but still, the differences are quite significant. 

Other parts that make the differences between platforms at the moment are:
- App launch time for UI Testing
    - takes 4-5 seconds, even with the simplest iOS apps
- Element selectors
    - e.g. scrolling to the element, even with the most optimised approaches
- Element assertions

I just wonder how these differences can be explained. How does having gray box testing potentially lead to better results in the above actions, and also in general?

**Reply by prolificcoder** [reply]
I am not exactly sure why xcuitest is slower but I know a person who might be able to answer @Shashikant86
